apps/extension/src/content-scripts/injected-penumbra-global.ts

```diff
@@ -5,54 +5,101 @@
-import { isPraxFailureMessageEvent, isPraxPortMessageEvent, PraxMessage } from './message-event';
+
+import {
+  isPraxFailureMessageEvent,
+  isPraxPortMessageEvent,
+  PraxMessage,
+  unwrapPraxMessage,
+} from './message-event';
 
 import { PraxConnection } from '../message/prax';
 
-const request = Promise.withResolvers<void>();
-
-// this is just withResolvers, plus a sync-queryable state attribute
-const connection = Object.assign(Promise.withResolvers<MessagePort>(), { state: false });
-connection.promise.then(
-  () => {
-    connection.state = true;
-    request.resolve();
-  },
-  () => {
-    connection.state = false;
-    request.reject();
-  },
-);
-
-// this resolves the connection promise when the isolated port script indicates
-const connectionListener = (msg: MessageEvent<unknown>) => {
-  if (isPraxPortMessageEvent(msg) && msg.origin === window.origin) {
-    // @ts-expect-error - ts can't understand the injected string
-    const praxPort: unknown = msg.data[PRAX];
-    if (praxPort instanceof MessagePort) connection.resolve(praxPort);
+type PromiseSettledResultStatus = PromiseSettledResult<unknown>['status'];
+
+class PraxInjection {
+  private static singleton?: PraxInjection = new PraxInjection();
+
+  public static get penumbra() {
+    return PraxInjection.singleton!.injection;
+  }
+
+  private manifestUrl = `${PRAX_ORIGIN}/manifest.json`;
+  private _request = Promise.withResolvers<void>();
+  private _connect = Promise.withResolvers<MessagePort>();
+  private _disconnect = Promise.withResolvers<void>();
+
+  private connectState?: PromiseSettledResultStatus;
+  private requestState?: PromiseSettledResultStatus;
+  private disconnectState?: PromiseSettledResultStatus;
+
+  private injection: Readonly<PenumbraInjection> = Object.freeze({
+    disconnect: () => this.endConnection(),
+    connect: () => (this.state() !== false ? this._connect.promise : this.connectionFailure),
+    isConnected: () => this.state(),
+    request: () => this.postRequest(),
+    manifest: String(this.manifestUrl),
+  });
+
+  private constructor() {
+    if (PraxInjection.singleton) return PraxInjection.singleton;
+
+    window.addEventListener('message', this.connectionListener);
+
+    void this._connect.promise
+      .then(
+        () => (this.connectState ??= 'fulfilled'),
+        () => (this.connectState ??= 'rejected'),
+      )
+      .finally(() => window.removeEventListener('message', this.connectionListener));
+
+    void this._disconnect.promise.then(
+      () => (this.disconnectState = 'fulfilled'),
+      () => (this.disconnectState = 'rejected'),
+    );
+
+    void this._request.promise.then(
+      () => (this.requestState = 'fulfilled'),
+      () => (this.requestState = 'rejected'),
+    );
+  }
+
+  private state(): boolean | undefined {
+    if (this.disconnectState) return false;
+    if (this.requestState === 'rejected') return false;
+    return this.connectState && this.connectState === 'fulfilled';
```

Comment:
It would be nice if the PraxInjection exposed state like:

```ts
enum ConnectionState {
  Idle = 'Idle',
  Connecting = 'Connecting',
  Connected = 'Connected',
  Disconnected = 'Disconnected',
  Disconnecting = 'Disconnecting',
  Failed = 'Failed',
}
```

Reply:
this is a private internal method that helps identify the appropriate boolean to return for the isConnected() interface.

i think adding an additional state variable that must be managed according to three other state variables is more complex than simply resolving the value upon demand.